### PR TITLE
[fix][broker] Fix MessageDeduplication replay timeout cause topic loading stuck

### PR DESCRIPTION
**Pull request by TakaHiR07**
Fixes https://github.com/apache/pulsar/issues/23003

### Motivation

As shown in the issue. fix the risk of topic loading stuck. 

### Modifications

1. make takeSnapshot() async , and if topic loading timeout, topic close should wait until takeSnapshot finish. This would not influence the other area where using takeSnapshot()
3. add test for this issue scene

### Verifying this change

- [x] Make sure that the change passes the CI checks.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/TopicDuplicationTest.java [hunk 1]
```diff
@@ -529,6 +538,89 @@ public void testDisableNamespacePolicyTakeSnapshotShouldNotThrowException() thro
         persistentTopic.checkDeduplicationSnapshot();
     }
 
+    @Test
+    public void testFinishTakeSnapshotWhenTopicLoading() throws Exception {
```
Comment (BewareMyPower) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/TopicDuplicationTest.java:542]: This test passed locally even without changes of `MessageDeduplication`, could you explain it?

image

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/TopicDuplicationTest.java [hunk 2]
```diff
@@ -529,6 +538,88 @@ public void testDisableNamespacePolicyTakeSnapshotShouldNotThrowException() thro
...
+        Field field2 = BrokerService.class.getDeclaredField("topics");
+        field2.setAccessible(true);
+        ConcurrentOpenHashMap<String, CompletableFuture<Optional<Topic>>> topics =
+                (ConcurrentOpenHashMap<String, CompletableFuture<Optional<Topic>>>)
+                        field2.get(pulsar.getBrokerService());
+
+        pulsar.getBrokerService().getTopic(topic, false);
+        Assert.assertTrue(topics.containsKey(topic));
+        CompletableFuture<Optional<Topic>> future = topics.get(topic);
+        future.completeExceptionally(FutureUtil.createTimeoutException("Failed to load topic within timeout",
```
Comment (shibd) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/TopicDuplicationTest.java:592]: https://github.com/apache/pulsar/blob/a91a172b4ee6d8b974a3fa905e435975557fcc57/pulsar-broker/src/test/java/org/apache/pulsar/client/api/OrphanPersistentTopicTest.java#L127-L137

Maybe we can refer to these code to mock topic load timeout and deduplication take snapshot timeout.

Comment (TakaHiR07) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/TopicDuplicationTest.java]: That's helpful. Thx. Have updated the test

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/TopicDuplicationTest.java (hunk 1 above)
Comment (TakaHiR07) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/TopicDuplicationTest.java:543]: Test has been improved. Currently would not passed

**Review by shibd — APPROVED** [review]

**Review by dao-jun — APPROVED** [review]

**Comment by lhotari** [comment]
@TakaHiR07 would you mind backporting this change to branch-3.2? The PR can be applied to branch-3.2, but the test TopicDuplicationTest.testFinishTakeSnapshotWhenTopicLoading fails consistently.

**Comment by TakaHiR07** [comment]
> @TakaHiR07 would you mind backporting this change to branch-3.2? The PR can be applied to branch-3.2, but the test TopicDuplicationTest.testFinishTakeSnapshotWhenTopicLoading fails consistently.

ok. I take some time to see the test.

**Comment by lhotari** [comment]
> @lhotari The reason is pr-22860 not in branch-3.2. Do I change the test "testFinishTakeSnapshotWhenTopicLoading" in branch-3.2, or wait until cherry-pick pr-22860?

Thanks for checking that @TakaHiR07 . I'll cherry-pick PR 22860 to branch-3.2 and check if things pass after that.

**Comment by lhotari** [comment]
@TakaHiR07 Cherry-picking PR 22860 fixed the issue. Thanks for the guidance.

**Pull request merged**